var/spack/repos/builtin/packages/kokkos/package.py

```diff
@@ -345,12 +354,26 @@ def cmake_args(self):
             if spec.variants[tpl].value:
                 options.append(self.define(tpl + "_DIR", spec[tpl].prefix))
 
-        if "+rocm" in self.spec:
-            options.append(self.define("CMAKE_CXX_COMPILER", self.spec["hip"].hipcc))
-        elif "+wrapper" in self.spec:
+        if "+wrapper" in self.spec:
             options.append(
                 self.define("CMAKE_CXX_COMPILER", self.spec["kokkos-nvcc-wrapper"].kokkos_cxx)
             )
+        elif "+rocm" in self.spec:
+            if "+compile_as_cmake_language" in self.spec:
+                options.append(
+                    self.define(
+                        "CMAKE_HIP_COMPILER",
+                        join_path(self.spec["llvm-amdgpu"].prefix.bin, "amdclang++"),
```

Comment:
Are we sure that the compiler is always `amdclang++`?

Reply:
`amdclang++` is the most likely case. It could potentially be just regular upstream clang if it was built with amdgpu target support. I believe this only becomes relevant once there is compiler-as-dependency support in Spack and if there is a way to define a `cuda_compiler` and `hip_compiler` somehow.

Reply:
Other occurrences of `CMAKE_HIP_COMPILER` use `clang++.` 

https://github.com/spack/spack/blob/9ac261af584c1ad6bdbcbad8b5ab1aec9283fa11/var/spack/repos/builtin/packages/py-onnxruntime/package.py#L201

https://github.com/spack/spack/blob/9ac261af584c1ad6bdbcbad8b5ab1aec9283fa11/lib/spack/spack/build_systems/cached_cmake.py#L285

Reply:
I've asked the AMD folks from the ElCapitan COE and they also say it should be `amdclang++`. @bvanessen @pranav-sivaraman do you know any particular reason why `clang++` was chosen in `CachedCMakePackage`?

Reply:
I'm not sure why `clang++` was chosen, but I do think `amdclang++` should be the default compiler to match the rest of Spack CI until a virtual `hip_compiler` gets added.

Reply:
I guess for now I would avoid using `cached_cmake.py` since the way to deal with HIP compilers will change in the future. I'm not sure if there are many packages are looking to compile with different HIP compilers so Kokkos should just implement it's own solution and it can be refactored some other day.

Reply:
As I see the message "This comment was marked as resolved." and the comment by @cedricchevalier19, I mark this conversation as resolved.